### PR TITLE
Added Integrator.setIntegrationForceGroups()

### PR DESCRIPTION
**Pull request by peastman**
See #2679.

### CODE REVIEW

File: openmmapi/include/openmm/Integrator.h [hunk 1]
```diff
@@ -85,6 +85,18 @@ class OPENMM_EXPORT Integrator {
      * @param steps   the number of time steps to take
      */
     virtual void step(int steps) = 0;
+    /**
+     * Get which force groups to use for integration.  By default, all force groups
+     * are included.  This is interpreted as a set if bit flags: the forces from group i
```
Comment (craabreu) [inline openmmapi/include/openmm/Integrator.h:90]: Small typo in "set of bit flags", both here and in line 96.

**Comment by craabreu** [comment]
@peastman Don't you believe that this new option should automatically change how potential energy is computed by calling `Context.getState(getPotentialEnergy=True)` without explicit group specification (and, consequently, the value reported by `StateDataReporter` with option `PotentialEnergy=True`)?

**Comment by peastman** [comment]
I'm not sure there's any good way of doing that.  The argument to `getState()` is an int containing bit flags.  The default value is 0xFFFFFFFF, which indicates a specific set of groups.  Likewise any other possible default value would indicate a particular set of groups.  How would you make the API work?

**Comment by craabreu** [comment]
How about simply moving the `setIntegrationForceGroups()` method from `Integrator` to `Context`? Maybe with another name such as `setActiveForceGroups()`, so that one of its purposes could be the definition of `getState()`'s default bit flag.

**Comment by peastman** [comment]
That wouldn't make any difference.  Currently `getState()` is declared like this:

```c++
State getState(int types, bool enforcePeriodicBox=false, int groups=0xFFFFFFFF) const;
```

You're suggesting to change it to something like this:

```c++
State getState(int types, bool enforcePeriodicBox=false, int groups=getIntegrator().getIntegrationForceGroups()) const;
```

But that isn't legal.  Default values for arguments aren't allowed to refer to the object the method is being called on.

**Comment by craabreu** [comment]
I was thinking of something like moving the `forceGroups` property from `Integrator` to `Context` and then doing this:

```cpp
State getState(int types, bool enforcePeriodicBox=false, int groups=-1) const;
```

Followed by replacing `groups` by `groups > 0 ? groups : forceGroups` in this line:

https://github.com/openmm/openmm/blob/b0d1358247cceee29e431d92224a9dc81d09b6b0/openmmapi/src/Context.cpp#L99

Something similar would have to be done for `ContextImpl.calcForcesAndEnergy()`. I believe that this would not be a real change in API, since backward compatibility seems to be fully preserved.

**Comment by peastman** [comment]
That would break backward compatibility by changing the meaning of `groups=0`.  Currently it means don't include any groups.

**Comment by craabreu** [comment]
> That would break backward compatibility by changing the meaning of groups=0. Currently it means don't include any groups.

Sorry. My mistake. The correct form would be `groups >= 0 ? groups : forceGroups`.

**Comment by peastman** [comment]
That still doesn't work.  It would mean that any time group 31 was included (hence the high order bit was set and the value was negative), it would ignore the groups they requested and instead return the default ones.

**Comment by z-gong** [comment]
Since the purpose is to calculate some user-defined 'force' or 'energy' without actually exerted on the system, could it be possible to add an attribute to the `Force` object instead of changing the logic of the integrator, barostat, etc.?

There could be an API like `force.setUsedInIntegration(False)` so that the force and energy evaluation will just return zero during normal integration.

An argument could be added to `getState` so that these forces are evaluated only when required, like `getState(includingUnusedForces=True)`

**Pull request merged**